kivy/core/camera/camera_avfoundation.pyx

```diff
@@ -66,6 +68,15 @@ class CameraAVFoundation(CameraBase):
         avf_camera_get_image(storage.camera,
             &width, &height, &rowsize, &data)
 
+        cyarr = cyview.array(
+            shape=(rowsize * height,),
+            itemsize=sizeof(char),
+            format="B",
+            mode="c",
+            allocate_buffer=False,
+        )
+        cyarr.data = data
```

Comment:
Absolutely OK for the `data == NULL` part, will do that ASAP.

By looking at the code, We could have a complete idea of the entire lifecycle of that pointer, cause We create and update it. ?And seems to be safe?

Reply:
I don't know the rules of the language of the .m files (I don't even know what language it is, presumably some c derivative). But I can see that the data pointer `data` is created with `malloc` in `updateImage` and that that buffer dies in the `~CameraFrame()` destructor.

Actually, when does that garbage collected? The image doesn't stay around forever, or does it? Is there maybe only one image and we simply pass around that one buffer? Or does a buffer get created for each image. If the former, then it's probably fine (although we need to be sure no one gets a hold of the buffer as it'll be overwritten by the next frame. If the latter, then we need to know how long the data pointer stays around?

Reply:
I guess the problem is that once we pass a pointer on to Python, we can't be sure how long a user may keep a hold of the pointer implicitly by keeping a ref to the cython array. So if they hold a ref to the array, but the camera has been shut off and the pointer deleted, trying to read the array value will access a dead pointer.

So we need to make sure to warn very clearly next to the `_buffer` variable that it cannot be used outside the camera internals so we keep full control over it. So maybe just add a comment next to the creation of `_buffer` saying it cannot be copied externally so one one makes this mistake?

Reply:
So, let's do a little brief:

The data buffer is created with a `malloc`over here:
https://github.com/kivy/kivy/blob/8aca43ed9322e1fe98ca231106691d77e4d684fd/kivy/core/camera/camera_avfoundation_implem.m#L644

The destructor of CameraFrame, where the buffer dies is called over here into the `dealloc` function of the `CaptureDelegate`:
https://github.com/kivy/kivy/blob/8aca43ed9322e1fe98ca231106691d77e4d684fd/kivy/core/camera/camera_avfoundation_implem.m#L577

Now, considering that the `release` function of a NSObject is something like that, the `dealloc` function is called when the object is released and the retainCount is equal to 0:
```
- (void)release
{
    _retainCount--;
    if (_retainCount == 0) {
        [self dealloc]
    }
}
``` 
The release function, for the `CaptureDelegate` is called over here, into the `Camera::stopCaptureDevice()` function:
https://github.com/kivy/kivy/blob/8aca43ed9322e1fe98ca231106691d77e4d684fd/kivy/core/camera/camera_avfoundation_implem.m#L271

The `stopCaptureDevice` function it's called into the `Camera::~Camera()`destructor, or into the `avf_camera_stop`function.

The `Camera::~Camera()`destructor, is called over here, by the `avf_camera_deinit` function, but  it's like a dead-end track, cause We don't use it :
https://github.com/kivy/kivy/blob/8aca43ed9322e1fe98ca231106691d77e4d684fd/kivy/core/camera/camera_avfoundation_implem.m#L689

The `avf_camera_stop` function is called from the `stop` function of the provider.

So, speaking about the `data` buffer, is available during the entire usage of the provider.

**_WARNING!!_ Not-so-fun fact**: While doing this whitepaper, to clarify the `data` buffer lifecycle, I found out, that the provider implementation doesn't implement a `__del__` function to deinit (and stop the provider if the user left that in `play`) when needed. I could also notice that into my notification bar (The green dot, in iOS > 14, it means that the Camera is in use) :
![IMG_0177](https://user-images.githubusercontent.com/8177736/103239120-f283d400-494c-11eb-9c5d-4f3dbcc1a895.jpg)

Reply:
Thanks for tracing this down.

About the `__del__`, yeah, that should be fixed (but can be in a different PR). But just be careful not do any python stuff from `__del__` because `__del__` can be called from any other part of the code in the middle of doing other stuff or even from other threads.

Reply:
Yes, let's do that into a different PR.
Will look at that during the next few days.